src/Tools/dotnet-monitor/CollectionRules/Actions/CollectGCDumpAction.cs

```diff
@@ -2,67 +2,114 @@
 using Microsoft.Diagnostics.Monitoring.WebApi;
 using Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Options.Actions;
 using System;
 using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
 using System.Threading;
 using System.Threading.Tasks;
 using Utils = Microsoft.Diagnostics.Monitoring.WebApi.Utilities;
-using System.ComponentModel.DataAnnotations;
 
 namespace Microsoft.Diagnostics.Tools.Monitor.CollectionRules.Actions
 {
-    internal sealed class CollectGCDumpAction : ICollectionRuleAction<CollectGCDumpOptions>
+    internal sealed class CollectGCDumpActionFactory :
+        ICollectionRuleActionFactory<CollectGCDumpOptions>
     {
         private readonly IServiceProvider _serviceProvider;
 
-        public CollectGCDumpAction(IServiceProvider serviceProvider)
+        public CollectGCDumpActionFactory(IServiceProvider serviceProvider)
         {
-            _serviceProvider = serviceProvider;
+            _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
         }
 
-        public async Task<CollectionRuleActionResult> ExecuteAsync(CollectGCDumpOptions options, IEndpointInfo endpointInfo, CancellationToken token)
+        public ICollectionRuleAction Create(IEndpointInfo endpointInfo, CollectGCDumpOptions options)
         {
-            if (options == null)
+            if (null == options)
             {
                 throw new ArgumentNullException(nameof(options));
             }
 
-            if (endpointInfo == null)
+            ValidationContext context = new(options, _serviceProvider, items: null);
+            Validator.ValidateObject(options, context, validateAllProperties: true);
+
+            return new CollectGCDumpAction(_serviceProvider, endpointInfo, options);
+        }
+
+        private sealed class CollectGCDumpAction :
+            ICollectionRuleAction,
+            IAsyncDisposable
+        {
+            private readonly CancellationTokenSource _disposalTokenSource = new();
+            private readonly IEndpointInfo _endpointInfo;
+            private readonly CollectGCDumpOptions _options;
+            private readonly IServiceProvider _serviceProvider;
+
+            private Task<CollectionRuleActionResult> _completionTask;
+
+            public CollectGCDumpAction(IServiceProvider serviceProvider, IEndpointInfo endpointInfo, CollectGCDumpOptions options)
             {
-                throw new ArgumentNullException(nameof(endpointInfo));
+                _endpointInfo = endpointInfo ?? throw new ArgumentNullException(nameof(endpointInfo));
+                _options = options ?? throw new ArgumentNullException(nameof(options));
+                _serviceProvider = serviceProvider ?? throw new ArgumentNullException(nameof(serviceProvider));
             }
 
-            string egress = options.Egress;
+            public async ValueTask DisposeAsync()
+            {
+                _disposalTokenSource.SafeCancel();
```

Comment:
Would it make sense to push some of this shared infrastructure around cleanup into a base class for actions?

Reply:
Should I go further than that and push all of DisposeAsync, StartAsync, and WaitForCompletion into a base class and have ExecuteAsync as an abstract method?

Reply:
I think that makes sense, there's a lot of shared functionality across these. We can always have multiple hierarchies if we find actions that greatly deviate in functionality.